File: src/Sequence.java
import java.util.ArrayList;


public class Sequence {
	public ArrayList<LoggedActivity> activity;
	private int sequenceId;
	private String groupingId; 
	
	public int getSequenceId() {
		return sequenceId;
	}
	
	public void setSequenceId(int sequenceId) {
		this.sequenceId = sequenceId;
	}
	
	public String getGroupingId() {
		return groupingId;
	}
	public void setGroupingId(String groupingId) {
		this.groupingId = groupingId;
	}
	
	public Sequence(int sequenceId) {
		super();
		this.sequenceId = sequenceId;
		this.activity = new ArrayList<LoggedActivity>();
		groupingId = "";
	}
	
	public String[] toString(String delimiter1, String delimiter2){
		String[] r = new String[2];
		r[0] = "";
		r[1] = "";
		for(LoggedActivity a : activity){
			r[0] += a.getActLabel() + delimiter1;
			r[1] += a.getActLabelReadable() + delimiter2;
		}
		if(r[0] != null && r[0].length()>0) r[0] = r[0].substring(0,r[0].length()-delimiter1.length());
		if(r[1] != null && r[1].length()>0) r[1] = r[1].substring(0,r[1].length()-delimiter2.length());
		return r;
	}
	
	public String toOutput(String delimiter, String actDelimiter){
		String r = sequenceId + delimiter + groupingId ;
		String l1 = "";
		String l2 = "";
		for(LoggedActivity a : activity){
			l1 += a.getActLabelReadable() + actDelimiter;
			l2 += a.getActLabel() + actDelimiter;
		}
		return r + delimiter + l1 + delimiter + l2;
	}
	
	public double totalTime(){
		double t = 0.0;
		if(activity != null) for(LoggedActivity a : activity){
			t += a.getTime();
		}
		return  t;
	}
	
	public int size(){
		if(activity != null) return activity.size();
		return 0;
	}
	
	public String getFirstDate(){
		if(activity != null && activity.size()>0) return activity.get(0).getDateStr();
		else return "";
	}
	public String getLastDate(){
		if(activity != null && activity.size()>0) return activity.get(activity.size()-1).getDateStr();
		else return "";
	}
}

File: src/models/Sequence.ts
import { LoggedActivity } from './LoggedActivity';

export class Sequence {
  activity: LoggedActivity[] = [];
  sequenceId: number;
  groupingId: string = '';

  constructor(sequenceId: number) {
    this.sequenceId = sequenceId;
    this.activity = [];
    this.groupingId = '';
  }

  toString(delimiter1: string, delimiter2: string): string[] {
    const r: string[] = ['', ''];
    for (const a of this.activity) {
      r[0] += a.actLabel + delimiter1;
      r[1] += a.actLabelReadable + delimiter2;
    }
    if (r[0].length > 0) {
      r[0] = r[0].substring(0, r[0].length - delimiter1.length);
    }
    if (r[1].length > 0) {
      r[1] = r[1].substring(0, r[1].length - delimiter2.length);
    }
    return r;
  }

  totalTime(): number {
    let t = 0.0;
    if (this.activity) {
      for (const a of this.activity) {
        t += a.time;
      }
    }
    return t;
  }

  size(): number {
    return this.activity ? this.activity.length : 0;
  }

  getFirstDate(): string {
    if (this.activity && this.activity.length > 0) {
      return this.activity[0].dateStr;
    }
    return '';
  }

  getLastDate(): string {
    if (this.activity && this.activity.length > 0) {
      return this.activity[this.activity.length - 1].dateStr;
    }
    return '';
  }
}

